### ISSUE TITLE
data_migrations table not populated via db:schema:load

### ISSUE THREAD

**Issue by tobyndockerill** [ilyakatz/data-migrate#65]
When starting an app with a blank database, the data_migrations table is not configured with the already-run data migrations. This is a known issue, but it seems as though a concrete use-case hasn't been mentioned before.

In the README there's a nice example of why this gem is useful. If I'm not mistaken, with this gem the workflow would look something like:

1. generate a schema migration to add a comment column to `Post`
2. generate a data migration to move the contents of the first comments to the `Post`
3. generate a schema migration to drop the `column_id` column from `Post`
4. generate a schema migration to drop the `comments` table

When running `rake db:migrate:with_data` the migrations would be combined and sorted, running them in the correct order and writing the timestamps to their respective `*_migrations` tables.

The problem comes when deploying the app to a new database. When running `rake db:schema:load` the migrations with timestamps before or equal to the timestamp version in `schema.rb` are inserted into the `schema_migrations` table (https://github.com/rails/rails/blob/4-2-stable/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L839 called from https://github.com/rails/rails/blob/4-2-stable/activerecord/lib/active_record/schema.rb#L45). After loading the schema, the `schema_migrations` table would have entries for migration numbers 1, 3, and 4, with the version number in `schema.rb` being the timestamp for migration 4. These migrations are then prevented from being unnecessarily re-run when `rake db:migrate` is run. 

There is no data migrations equivalent.

This means that running `rake db:migrate:with_data` after the schema is loaded will cause the data migration (migration 2 above) to be executed, because there is no corresponding entry in the `data_migrations` table. This will raise an exception because the comments table does not exist anymore.

@cheerfulstoic initially raised the issue (#2) and @edk proposed a fix (#20).

There is a small edge-case with @edk's solution and that is it always assumes that a schema migration is always the last migration. If a data migration comes last, it would be re-run.

A solution that fixes this edge case would be for gem to create its own schema file where it would keep track of the current version.
```
# db/data_schema.rb

ActiveRecord::Schema.define(version: 20170807000000) do; end
```
Then, attempting to keep as many similarities to the way Rails does it, we would update this version every time a data migration is successfully completed and use it to determine which data migrations have already been completed.

Are there any misconceptions I've made about the gem's behaviour?
Is this proposed solution the best one?

I'm happy to discuss here and submit a PR if we're agreed that this is an issue that should be fixed.

**Comment by ilyakatz** [comment]
That's a very interesting issue. So, if I understand you correctly, you want 

1. to load a brand new database with `rake db:schema:load`
1. then run `rake data:migrate` (or `rake db:migrate:with_data` which will only run data migrations since all schema migrations have been completed). 

This will fail because some schema changes made previous data migration obsolete. 

I don't think I have an answer for this. If we go by your suggestion `db/data_schema.rb`, will this mean that it will skip all previous data migrations? If so, this will make the task of repeating history (full schema + data migrations in the order they were created) impossible.

**Comment by ilyakatz** [comment]
On second thought, perhaps what you want to say is this "I'm going to run `rake db:schema:load` so mark all data migrations up to now as **skipped**"?

**Comment by tobyndockerill** [comment]
To address the issue of repeating history, I'm not sure if that's possible with plain schema migrations anyway. If our aim is to be able to run these schema migrations on a blank database, we first need to set up the schema, which will automatically enter the `schema_migration` timestamps. From that point on (assuming there are no new schema migrations) a `rake db:migrate` won't actually do anything.

It could be an interesting possibility to mark certain _data_ migrations as skipped like you suggest, but this would diverge from the way schema migrations are handled.

**Comment by ilyakatz** [comment]
Right, repeating history will only work if you go through each individual db/migration + data/migration. Which has been very useful.

You are correct, skipping would diverge from regular rails. Let me think about this some more. 

FYI: There is a rake task `rake data:forward` which is _supposed_ to have done exactly that but seems it hasn't worked for years 😆

**Comment by tobyndockerill** [comment]
I just thought I'd give a little more detail about my workflow and why this has become an issue for me.

The app I work on is quite mature and has had many migrations over its lifetime. New instances are deployed often and running each and every migration from the beginning is just not possible as it would take far too long. 

The alternative is to run a `schema:load` to set the database up without having to run through all prior migrations. This is where the automatic insertion into the `schema_migrations` table happens, and it seems as though this is the Rails-recommended way (see the comment automatically added to `schema.rb`: https://github.com/rails/rails/blob/da589fcc4a15cdc923b9df55f9fe964df996e49b/activerecord/lib/active_record/schema_dumper.rb#L70). 

To provide the same functionality for data migrations as ActiveRecord does for schema migrations, I'm going to submit a PR within the next 12 hours or so that will do the following things:
- add `db:schema:load:with_data` which will call `db:schema:load`, then read the version in `db/data_schema.rb` and insert the timestamps of data migrations that we assume to have already run based on that version
- add `data:dump` which will write the current data version to `db/data_schema.rb`. This will be run every time `db:_dump` is called in addition to when specific data migrations are run, such as `data:migrate` or `data:rollback`

This will allow us to set up a new instance without needing to run migrations (data or schema) with a single idiomatic rake task (with all concepts borrowed from existing ActiveRecord functionality).

**Comment by ilyakatz** [comment]
@tobyndockerill now that i read through your workflow, it makes more sense. let me play around with this a bit more and and if we do decide to proceed with this, I do a review of the PR, sounds good?

**Comment by ilyakatz** [comment]
also, i will be traveling for 2 weeks, so won't be able to get to it right away but it's on my radar!

**Comment by tobyndockerill** [comment]
No worries at all, that's perfect. Thanks for taking the time to look into it.

**Comment by philipedwards** [comment]
Our organisation has the same workflow outlined by @tobyndockerill. We use schema:load to populate our database as applying migrations from the beginning of time is not feasible. 

With the data migrations table empty after a schema load, data-migrate attempts to run all data migrations. This fails as the database isn't in sync with the migrations. 

We've been using the code from the PR listed above and can confirm that it solves the problem. I also like that the code closely resembles/is structured similarly as the Rails code that handles dumping the DB schema version.

+1

**Comment by tobyndockerill** [comment]
Hi @ilyakatz, I was wondering if you've had a chance to review my PR addressing this issue?

**Comment by bbugh** [comment]
We were about to consider using this (awesome) gem but the lack of support for `db:schema:load` means that our workflow isn't going to flow very well. We have a complex data migration and boy am I ever tired of partially migrating and running rake tasks instead of it just being automated and tested.

Have y'all come up with a workaround for this until #68 has been accepted?

**Comment by ilyakatz** [comment]
Hi @bbugh, I tested out @tobyndockerill's solution a but and it seems to do the job. I made some comments on the PR which are mostly related to keeping things more maintainable not the actual functionality. So if you want you can set you gemfile to use his repo/branch and then when it's merged you can set it back to the main gem from rubygems.

Obviously, I would recommend caution and testing things out before using it in production. You can also provide some feedback about how this works out ;)

**Comment by tobyndockerill** [comment]
I've been using my solution in production over the last month or so and it's been working well. I've been super busy lately but implementing the changes on the PR is still on my list of things to do! I'll most probably get around to it this weekend.

**Comment by ilyakatz** [comment]
This is now available via #75 

thanks @tobyndockerill
